src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE DeriveAnyClass #-}
+
+module Occurrence where
+
+import Prelude hiding (abs, iterate, max, sum)
+import Prelude qualified
+
+import Control.Monad.State.Strict
+import Data.IntMap.Strict qualified as M
+
+import Name
+
+-- === Use Analysis ===
+
+-- We want to know how a binding is used in a Dex program, so as to
+-- know whether inlining it is guaranteed to conserve work and code
+-- size.
+
+-- Conservation of code size is easy.  Coservation of work, however,
+-- is more subtle.  Inlining the computation of x into
+--   for i. x
+-- does not conserve work (unless the index set has size 0 or 1), but
+-- inlining x into
+--   for i. x.i
+-- does conserve work -- provided we have access to the loop body
+-- that constructs it so we can beta-reduce against the indexing
+-- operation.
+
+-- The main difficulty is that Dex allows indexing by expressions
+-- other than referencing a loop binder, so to solve subtle cases we
+-- need to translate iteration-space expressions into data-space.
+-- For example, inlining x into
+--   for i. x.(Left i)
+--   for j. x.(Right j)
+-- conserves work, because the two injections have disjoint ranges.
```

Comment:
Note that this depends on the Ix instance. I think that an `Ix ((Fin n)|(Fin n))` that collapses both cases onto a single `Fin n` is lawful (`size = n` and let's say that `unsafe_from_ordinal` always picks the left case), but would not be work-preserving in this case!

Reply:
I was under the impression that we expect every `Ix n` instance to implement a bijection between `n` and a prefix of the non-negative integers.  Is that not so?

Reply:
Good question. Technically, it would still work even if it was a surjection onto the prefix given by the size. So it's really a question of the contract we pick for `Ix`.